src/jws/verify.rs

```diff
@@ -0,0 +1,544 @@
+use askar_crypto::sign::KeySigVerify;
+
+use crate::{
+    error::{err_msg, ErrorKind, Result, ResultExt},
+    jws::ParsedJWS,
+};
+
+impl<'a, 'b> ParsedJWS<'a, 'b> {
+    pub(crate) fn verify<Key: KeySigVerify>(&self, signer: (&str, &Key)) -> Result<bool> {
+        let (kid, key) = signer;
+
+        let (i, signature) = self
+            .jws
+            .signatures
+            .iter()
+            .enumerate()
+            .find(|(_, sig)| sig.header.kid == kid)
+            .ok_or_else(|| err_msg(ErrorKind::InvalidState, "KID not found"))?;
+
+        let protected = self
+            .protected
+            .get(i)
+            .ok_or_else(|| err_msg(ErrorKind::InvalidState, "Invalid protected header index"))?;
+
+        let sig_type = protected.alg.sig_type()?;
+        let sign_input = format!("{}.{}", signature.protected, self.jws.payload);
+
+        let signature = base64::decode_config(&signature.signature, base64::URL_SAFE_NO_PAD)
+            .kind(ErrorKind::Malformed, "Unable decode signature")?;
+
+        let valid = key
+            .verify_signature(sign_input.as_bytes(), &signature, Some(sig_type))
+            .kind(ErrorKind::Malformed, "Unable verify signature")?;
+
+        Ok(valid)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use askar_crypto::{
+        alg::{ed25519::Ed25519KeyPair, k256::K256KeyPair, p256::P256KeyPair},
+        jwk::FromJwk,
+        sign::KeySigVerify,
+    };
+
+    use crate::{
+        error::{Error, ErrorKind},
+        jws,
+    };
+
+    #[test]
+    fn verify_works() {
+        _verify_works::<Ed25519KeyPair>(ALICE_KID_ED25519, ALICE_PKEY_ED25519, ALICE_MSG_ED25519);
+        _verify_works::<P256KeyPair>(ALICE_KID_P256, ALICE_PKEY_P256, ALICE_MSG_P256);
+        _verify_works::<K256KeyPair>(ALICE_KID_K256, ALICE_PKEY_K256, ALICE_MSG_K256);
+
+        fn _verify_works<K: FromJwk + KeySigVerify>(kid: &str, key: &str, msg: &str) {
+            let res = _verify::<K>(kid, key, msg);
+            let res = res.expect("res is err");
+            assert_eq!(res, true);
+        }
+    }
+
+    #[test]
+    fn verify_works_multiple_signatures() {
```

Comment:
What if one signature is valid, while another is invalid?

Reply:
The answer is on higher level. `verify` allows to check only one signature.